core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -110,16 +114,60 @@ static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProv
                 .whereIsInstanceOfAny(JoinNode.class, UnnestNode.class)
                 .matches();
         if (hasExpandingNodes) {
-            return Double.NaN;
+            return NaN;
         }
 
         List<PlanNode> sourceNodes = PlanNodeSearcher.searchFrom(node, lookup)
                 .whereIsInstanceOfAny(TableScanNode.class, ValuesNode.class)
                 .findAll();
 
-        return sourceNodes.stream()
+        double sourceTablesSizeInBytes = sourceNodes.stream()
                 .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
                 .sum();
+
+        double bestKnownRowCount = getBestKnownRowCount(node, lookup, statsProvider);
+        double sourceTablesRowCount = sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputRowCount())
+                .sum();
+
+        if (bestKnownRowCount < sourceTablesRowCount) {
+            return (bestKnownRowCount / sourceTablesRowCount) * sourceTablesSizeInBytes;
```

Comment:
This looks a bit like a second derivative estimate (e.g. we use estimated `getBestKnownRowCount` to calculate final estimate based on `sourceTablesRowCount`)

Why not to use `output size in bytes` from first available plan nodes that provide it (when running recursion)?

Reply:
I avoided using `output size in bytes` directly because I was concerned about underestimation in cases where the symbol stats might be missing for some columns (e.g. due to unestimated terms like aggregates in AggregationStatsRule).